In [2]:
import pandas as pd
from pandas import Series, DataFrame

In [ ]:
# to write a summary on this note 

In [94]:
# exercise 
filename = r'C:\Users\dqthi\OneDrive\Coding_Journey\CodeBreakThrough\Beginner_Python_Programming_All-in-One_Tutorial_Series\Pandas\University_exercise\college.csv'

df = pd.read_csv(filename, index_col = "Unnamed: 0")

In [95]:
df

CITY STABBR  SAT_AVG  \
0               Normal     AL    929.0   
1           Birmingham     AL   1195.0   
2           Montgomery     AL      NaN   
3           Huntsville     AL   1322.0   
4           Montgomery     AL    935.0   
...                ...    ...      ...   
7107              Yuma     AZ      NaN   
7108          San Jose     CA      NaN   
7109         Lancaster     CA      NaN   
7110            Malden     MA      NaN   
7111  Lauderdale Lakes     FL      NaN   

                                                 INSTNM  ADM_RATE  MTHCMP1  \
0                              Alabama A & M University    0.9027      NaN   
1                   University of Alabama at Birmingham    0.9181      NaN   
2                                    Amridge University       NaN      NaN   
3                   University of Alabama in Huntsville    0.8123      NaN   
4                              Alabama State University    0.9787      NaN   
...                                                 ...       ...      ...   
7107                   Arizona State University at Yuma       NaN      NaN   
7108  Bay Area Medical Academy - San Jose Satellite ...       NaN      9.0   
7109                        High Desert Medical College       NaN     21.0   
7110                                       BCI - Malden       NaN     10.0   
7111          Health Career Institute- Lauderdale Lakes       NaN      NaN   

      TUITIONFEE_IN  
0            9857.0  
1            8328.0  
2            6900.0  
3           10280.0  
4           11068.0  
...             ...  
7107            NaN  
7108            NaN  
7109            NaN  
7110            NaN  
7111        22575.0  

[7112 rows x 7 columns]

In [96]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7112 entries, 0 to 7111
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   CITY           7112 non-null   object 
 1   STABBR         7112 non-null   object 
 2   SAT_AVG        1320 non-null   float64
 3   INSTNM         7112 non-null   object 
 4   ADM_RATE       2037 non-null   float64
 5   MTHCMP1        2528 non-null   float64
 6   TUITIONFEE_IN  4021 non-null   float64
dtypes: float64(4), object(3)
memory usage: 444.5+ KB


observation: 
1) Sat column is missing alot of data. Only 1320 rows have data, compared to the total rows of 7112
2) 3 other columns that have missing data are: a) adm_rate (2037), Average month to complete (2528), tuition fee for in-states
3) index column is the number of the row 


In [102]:
len(df['STABBR'].value_counts().index)

59

In [98]:
df_index_state_city = df.set_index(['STABBR','CITY'])

In [34]:
df_index_state_city.loc['MA']['SAT_AVG'].mean() # this is to use the .loc() to step 1: getting the mean value of column 'SAT_AVG' for index value of "MA"

1209.3191489361702

In [38]:
df_index_state_city.loc['IL']['ADM_RATE'].mean()*100

67.04847222222222

In [ ]:
# the acceptance rate of IL is 67%, which mean that out of 100 application,, 67 will be accepted to the schools 

In [19]:
df['SAT_AVG'][df['STABBR'] == ('MA')].mean()  # this approach is to search for column State with value of "MA", then calculate the mean for column 'SAT average' 

1209.3191489361702

In [41]:
df_index_city_state = df_index_state_city.swaplevel()

In [42]:
df_index_city_state 

Unnamed: 0  SAT_AVG  \
CITY             STABBR                        
Normal           AL               0    929.0   
Birmingham       AL               1   1195.0   
Montgomery       AL               2      NaN   
Huntsville       AL               3   1322.0   
Montgomery       AL               4    935.0   
...                             ...      ...   
Yuma             AZ            7107      NaN   
San Jose         CA            7108      NaN   
Lancaster        CA            7109      NaN   
Malden           MA            7110      NaN   
Lauderdale Lakes FL            7111      NaN   

                                                                    INSTNM  \
CITY             STABBR                                                      
Normal           AL                               Alabama A & M University   
Birmingham       AL                    University of Alabama at Birmingham   
Montgomery       AL                                     Amridge University   
Huntsville       AL                    University of Alabama in Huntsville   
Montgomery       AL                               Alabama State University   
...                                                                    ...   
Yuma             AZ                       Arizona State University at Yuma   
San Jose         CA      Bay Area Medical Academy - San Jose Satellite ...   
Lancaster        CA                            High Desert Medical College   
Malden           MA                                           BCI - Malden   
Lauderdale Lakes FL              Health Career Institute- Lauderdale Lakes   

                         ADM_RATE  MTHCMP1  TUITIONFEE_IN  
CITY             STABBR                                    
Normal           AL        0.9027      NaN         9857.0  
Birmingham       AL        0.9181      NaN         8328.0  
Montgomery       AL           NaN      NaN         6900.0  
Huntsville       AL        0.8123      NaN        10280.0  
Montgomery       AL        0.9787      NaN        11068.0  
...                           ...      ...            ...  
Yuma             AZ           NaN      NaN            NaN  
San Jose         CA           NaN      9.0            NaN  
Lancaster        CA           NaN     21.0            NaN  
Malden           MA           NaN     10.0            NaN  
Lauderdale Lakes FL           NaN      NaN        22575.0  

[7112 rows x 6 columns]

In [53]:
df_index_city_state.loc['Cambridge']['INSTNM'].count()

6

In [56]:
df_index_city_state.loc['Cambridge']['SAT_AVG'].mean()

1392.6666666666667

In [ ]:
# there are 6 schools in the city named Cambridge 
# the average SAT scores for school in Cambridge is much higher than the average of the nation, which is 1392.6 


In [75]:
df['STABBR'] == ('CA' | 'IL' | 'NY')

TypeError: unsupported operand type(s) for |: 'str' and 'str'

In [87]:
df_ca_il_ny = df_index_state_city.loc[['CA', 'IL', 'NY']]


In [92]:
df_ca_il_ny.loc['IL'].sort_values('ADM_RATE', ascending=False)

Unnamed: 0  SAT_AVG                                        INSTNM  \
CITY                                                                            
Peoria            1011      NaN                              Midstate College   
Chicago           1073   1215.0                   VanderCook College of Music   
Skokie            5137      NaN                       Estelle Medical Academy   
Chicago           1026      NaN           Northwestern College-Chicago Campus   
Skokie             962   1064.0                    Hebrew Theological College   
...                ...      ...                                           ...   
Belleville        6875      NaN              Lindenwood University-Belleville   
Aurora            7029      NaN                      Rasmussen College-Aurora   
Romeoville        7030      NaN           Rasmussen College-Romeoville/Joliet   
Mokena            7031      NaN          Rasmussen College-Mokena/Tinley Park   
Chicago           7087      NaN  Relay Graduate School of Education - Chicago   

            ADM_RATE  MTHCMP1  TUITIONFEE_IN Stabbr  
CITY                                                 
Peoria        1.0000      NaN        16230.0     CA  
Chicago       1.0000      NaN        28434.0     CA  
Skokie        1.0000      NaN         8680.0     CA  
Chicago       0.9615      NaN        17475.0     CA  
Skokie        0.9500      NaN        17630.0     CA  
...              ...      ...            ...    ...  
Belleville       NaN      NaN        16960.0     CA  
Aurora           NaN      NaN        10935.0     CA  
Romeoville       NaN      NaN        10935.0     CA  
Mokena           NaN      NaN        10935.0     CA  
Chicago          NaN      NaN            NaN     CA  

[277 rows x 7 columns]

In [ ]:
# three schools in IL that have the highest admission rate are: Midstate College, VanderCOok College of Music, Estelle Medical Academy 

In [93]:
df_ca_il_ny.loc['CA'].sort_values('ADM_RATE')

Unnamed: 0  SAT_AVG  \
CITY                              
Stanford          3979   1489.0   
Pasadena           212   1558.0   
Claremont          427   1461.0   
Claremont          268   1452.0   
Claremont          325   1527.0   
...                ...      ...   
Fremont           7081      NaN   
Merced            7095      NaN   
Emeryville        7097      NaN   
San Jose          7108      NaN   
Lancaster         7109      NaN   

                                                       INSTNM  ADM_RATE  \
CITY                                                                      
Stanford                                  Stanford University    0.0473   
Pasadena                   California Institute of Technology    0.0774   
Claremont                                      Pomona College    0.0836   
Claremont                           Claremont McKenna College    0.1036   
Claremont                                 Harvey Mudd College    0.1542   
...                                                       ...       ...   
Fremont                                        Unitek College       NaN   
Merced                               WestMed College - Merced       NaN   
Emeryville          SAE Institute of Technology San Francisco       NaN   
San Jose    Bay Area Medical Academy - San Jose Satellite ...       NaN   
Lancaster                         High Desert Medical College       NaN   

            MTHCMP1  TUITIONFEE_IN Stabbr  
CITY                                       
Stanford        NaN        49617.0     CA  
Pasadena        NaN        49908.0     CA  
Claremont       NaN        51075.0     CA  
Claremont       NaN        52825.0     CA  
Claremont       NaN        54636.0     CA  
...             ...            ...    ...  
Fremont         NaN        35070.0     CA  
Merced         14.0            NaN     CA  
Emeryville      9.0            NaN     CA  
San Jose        9.0            NaN     CA  
Lancaster      21.0            NaN     CA  

[716 rows x 7 columns]

In [ ]:
# Apparently Stanford University is the one that has lowest accepatance rate of 4.73%  